File: shared/treeLayout.ts
/**
 * Family-tree layout — pure geometry, no rendering. Assigns each reachable person a
 * generation (0 = focus, negative = ancestors, positive = descendants) and an (x, y).
 *
 * Couples are handled with care for the messy real cases: spouses AND unmarried
 * co-parents (two people sharing a child) are grouped ADJACENT; within a couple the
 * man goes left and the woman right (same-sex couples order by birth year, then id);
 * a person married more than once appears ONCE with their spouses chained beside
 * them. Pedigree collapse keeps one node per person even when reachable by several
 * paths. Each node carries `coupleSide` so the renderer can face default portraits
 * inward. The view maps ids back to Person records; this module only needs edges +
 * light attributes, so it stays testable.
 */

export type CoupleSide = 'left' | 'right' | 'none';

export interface TreePersonAttr {
  id: string;
  sex?: string;
  birthYear?: number | null;
}

export interface TreeNode {
  personId: string;
  generation: number;
  x: number;
  y: number;
  coupleSide: CoupleSide;
}

export interface TreeEdge {
  from: string;
  to: string;
  kind: 'parent' | 'spouse';
}

export interface TreeLayoutInput {
  focusId: string;
  parentEdges: { parent: string; child: string }[];
  spouseEdges: { a: string; b: string }[];
  persons?: TreePersonAttr[];
  ancestorDepth?: number;
  descendantDepth?: number;
  nodeWidth?: number;
  nodeHeight?: number;
  hGap?: number;
  vGap?: number;
}

export interface TreeLayoutResult {
  nodes: TreeNode[];
  edges: TreeEdge[];
  width: number;
  height: number;
}

// Ancestor/descendant depth is UNLIMITED by default — a family tree should show every
// generation it has, however deep. A caller may still pass a finite depth to focus.
const DEFAULTS = {
  ancestorDepth: Number.POSITIVE_INFINITY,
  descendantDepth: Number.POSITIVE_INFINITY,
  nodeWidth: 160,
  nodeHeight: 64,
  hGap: 28,
  vGap: 88,
};

function pairKey(a: string, b: string): string {
  return a < b ? `${a}|${b}` : `${b}|${a}`;
}

export function computeTreeLayout(input: TreeLayoutInput): TreeLayoutResult {
  const opts = { ...DEFAULTS, ...input };
  if (!input.focusId) return { nodes: [], edges: [], width: 0, height: 0 };

  const sexOf = new Map<string, string>();
  const birthYearOf = new Map<string, number | null>();
  for (const p of input.persons ?? []) {
    sexOf.set(p.id, p.sex ?? 'unknown');
    birthYearOf.set(p.id, p.birthYear ?? null);
  }

  const parentsOf = new Map<string, string[]>();
  const childrenOf = new Map<string, string[]>();
  for (const { parent, child } of input.parentEdges) {
    (childrenOf.get(parent) ?? childrenOf.set(parent, []).get(parent)!).push(child);
    (parentsOf.get(child) ?? parentsOf.set(child, []).get(child)!).push(parent);
  }
  const spousesOf = new Map<string, string[]>();
  for (const { a, b } of input.spouseEdges) {
    (spousesOf.get(a) ?? spousesOf.set(a, []).get(a)!).push(b);
    (spousesOf.get(b) ?? spousesOf.set(b, []).get(b)!).push(a);
  }

  // ── Generations (BFS up + down; nearest-to-focus wins; spouses share a gen) ──
  const generation = new Map<string, number>();
  generation.set(input.focusId, 0);
  const queue = [input.focusId];
  while (queue.length) {
    const id = queue.shift()!;
    const g = generation.get(id)!;
    if (g > -opts.ancestorDepth) {
      for (const parent of parentsOf.get(id) ?? []) {
        if (!generation.has(parent)) {
          generation.set(parent, g - 1);
          queue.push(parent);
        }
      }
    }
    if (g < opts.descendantDepth) {
      for (const childId of childrenOf.get(id) ?? []) {
        if (!generation.has(childId)) {
          generation.set(childId, g + 1);
          queue.push(childId);
        }
      }
    }
    for (const spouse of spousesOf.get(id) ?? []) {
      if (!generation.has(spouse) && Math.abs(g) <= Math.max(opts.ancestorDepth, opts.descendantDepth)) {
        generation.set(spouse, g);
        queue.push(spouse);
      }
    }
  }

  const present = new Set(generation.keys());

  // ── Couple links within a generation: spouses + unmarried co-parents ────────
  const coupleLinks = new Set<string>(); // "a|b"
  const partners = new Map<string, Set<string>>();
  const link = (a: string, b: string) => {
    if (a === b || !present.has(a) || !present.has(b) || generation.get(a) !== generation.get(b)) return;
    coupleLinks.add(pairKey(a, b));
    (partners.get(a) ?? partners.set(a, new Set()).get(a)!).add(b);
    (partners.get(b) ?? partners.set(b, new Set()).get(b)!).add(a);
  };
  for (const { a, b } of input.spouseEdges) link(a, b);
  // Co-parents: any two parents of the same child.
  for (const parents of parentsOf.values()) {
    for (let i = 0; i < parents.length; i++) {
      for (let j = i + 1; j < parents.length; j++) link(parents[i], parents[j]);
    }
  }

  const byGen = new Map<number, string[]>();
  for (const [id, g] of generation) (byGen.get(g) ?? byGen.set(g, []).get(g)!).push(id);
  const gens = [...byGen.keys()].sort((a, b) => a - b);

  // Order a couple (2 people): man left, woman right; same-sex by birth year then id.
  const orderPair = (x: string, y: string): [string, string] => {
    const sx = sexOf.get(x);
    const sy = sexOf.get(y);
    if (sx === 'male' && sy === 'female') return [x, y];
    if (sx === 'female' && sy === 'male') return [y, x];
    const bx = birthYearOf.get(x) ?? null;
    const by = birthYearOf.get(y) ?? null;
    if (bx != null && by != null && bx !== by) return bx < by ? [x, y] : [y, x];
    return x < y ? [x, y] : [y, x];
  };

  // Order a whole coupled component into a contiguous run (person + chained spouses).
  const orderComponent = (members: string[]): string[] => {
    if (members.length === 1) return members;
    if (members.length === 2) return orderPair(members[0], members[1]);
    // Chain/star: greedy walk from an endpoint, then man-left tidy for the first pair.
    const set = new Set(members);
    const start = members.find((m) => [...(partners.get(m) ?? [])].filter((p) => set.has(p)).length === 1) ?? members[0];
    const seq: string[] = [];
    const seen = new Set<string>();
    let cur: string | undefined = start;
    while (cur) {
      seq.push(cur);
      seen.add(cur);
      const next: string | undefined = [...(partners.get(cur) ?? [])].find((p) => set.has(p) && !seen.has(p));
      cur = next;
    }
    for (const m of members) if (!seen.has(m)) seq.push(m); // stragglers (branches)
    return seq;
  };

  const order = new Map<string, number>();

  const orderGeneration = (g: number, neighbourGen: number | null) => {
    const ids = byGen.get(g)!;
    // Components via union of coupleLinks restricted to this generation.
    const parent = new Map<string, string>();
    const find = (x: string): string => {
      let r = x;
      while (parent.get(r) !== r) r = parent.get(r)!;
      return r;
    };
    for (const id of ids) parent.set(id, id);
    for (const id of ids) {
      for (const p of partners.get(id) ?? []) {
        if (generation.get(p) === g) parent.set(find(id), find(p));
      }
    }
    const comps = new Map<string, string[]>();
    for (const id of ids) (comps.get(find(id)) ?? comps.set(find(id), []).get(find(id))!).push(id);

    const nodeBary = (id: string): number => {
      if (neighbourGen === null) return Number.MAX_SAFE_INTEGER;
      const rel = g < 0 ? childrenOf.get(id) ?? [] : parentsOf.get(id) ?? [];
      const placed = rel.filter((n) => generation.get(n) === neighbourGen && order.has(n)).map((n) => order.get(n)!);
      return placed.length ? placed.reduce((s, v) => s + v, 0) / placed.length : Number.MAX_SAFE_INTEGER;
    };

    const ordered = [...comps.values()]
      .map((members) => {
        const seq = orderComponent(members);
        const barys = members.map(nodeBary).filter((b) => b < Number.MAX_SAFE_INTEGER);
        const bary = barys.length ? barys.reduce((s, v) => s + v, 0) / barys.length : Number.MAX_SAFE_INTEGER;
        return { seq, bary };
      })
      .sort((a, b) => a.bary - b.bary);

    let i = 0;
    for (const comp of ordered) for (const id of comp.seq) order.set(id, i++);
  };

  // Seed the focus generation, then order outward so couples sit under their kin.
  orderGeneration(0, null);
  for (const g of [...gens].sort((a, b) => Math.abs(a) - Math.abs(b))) {
    if (g === 0) continue;
    orderGeneration(g, g < 0 ? g + 1 : g - 1);
  }

  // ── Coordinates ─────────────────────────────────────────────────────────────
  const minGen = gens[0] ?? 0;
  const rowStep = opts.nodeHeight + opts.vGap;
  const colStep = opts.nodeWidth + opts.hGap;
  const nodesById = new Map<string, TreeNode>();
  const nodes: TreeNode[] = [];
  let maxCols = 0;
  for (const g of gens) {
    const ids = byGen.get(g)!.slice().sort((a, b) => (order.get(a) ?? 0) - (order.get(b) ?? 0));
    maxCols = Math.max(maxCols, ids.length);
    ids.forEach((id, col) => {
      const node: TreeNode = { personId: id, generation: g, x: col * colStep, y: (g - minGen) * rowStep, coupleSide: 'none' };
      nodes.push(node);
      nodesById.set(id, node);
    });
  }

  // coupleSide: relative to the nearest placed partner (drives inward-facing portraits).
  for (const node of nodes) {
    const ps = [...(partners.get(node.personId) ?? [])].map((p) => nodesById.get(p)).filter(Boolean) as TreeNode[];
    if (ps.length === 0) continue;
    const nearest = ps.reduce((a, b) => (Math.abs(b.x - node.x) < Math.abs(a.x - node.x) ? b : a));
    node.coupleSide = node.x < nearest.x ? 'left' : 'right';
  }

  const edges: TreeEdge[] = [];
  for (const { parent, child } of input.parentEdges) {
    if (present.has(parent) && present.has(child)) edges.push({ from: parent, to: child, kind: 'parent' });
  }
  for (const { a, b } of input.spouseEdges) {
    if (present.has(a) && present.has(b)) edges.push({ from: a, to: b, kind: 'spouse' });
  }

  return {
    nodes,
    edges,
    width: Math.max(0, maxCols * colStep - opts.hGap),
    height: Math.max(0, gens.length * rowStep - opts.vGap),
  };
}

File: ios/Packages/NodusKit/Sources/NodusKit/Genealogy/TreeLayout.swift
import Foundation

/// Family-tree layout — pure geometry, no rendering.
///
/// A port of `shared/treeLayout.ts`, kept faithful because the desktop's version already
/// solves the messy real cases and a second, subtly different arrangement of the same family
/// would be worse than no tree at all:
///
/// - spouses **and** unmarried co-parents sit adjacent, because sharing a child is what makes
///   two people a couple on a chart;
/// - within a couple the man goes left and the woman right, same-sex couples by birth year
///   then id, so the layout is stable rather than dependent on row order;
/// - somebody married more than once appears **once**, with their spouses chained beside them;
/// - pedigree collapse keeps one node per person even when they are reachable by several paths
///   — which happens in every real family tree that goes back far enough.
public enum TreeLayout {
    public enum CoupleSide: String, Sendable, Hashable { case left, right, none }

    public enum Branch: String, Sendable, Hashable, Codable {
        case paternal, maternal, neutral
    }

    public struct PersonAttributes: Sendable, Hashable {
        public let id: String
        public let sex: String?
        public let birthYear: Int?

        public init(id: String, sex: String?, birthYear: Int?) {
            self.id = id
            self.sex = sex
            self.birthYear = birthYear
        }
    }

    public struct Node: Sendable, Hashable, Identifiable {
        public let personId: String
        /// 0 = focus, negative = ancestors, positive = descendants.
        public let generation: Int
        public var x: Double
        public var y: Double
        public var coupleSide: CoupleSide
        public var branch: Branch
        /// 0 at the focus's own generation, growing with distance — drives the shade.
        public var tone: Double

        public var id: String { personId }
    }

    public struct Edge: Sendable, Hashable {
        public enum Kind: String, Sendable { case parent, spouse, sibling }
        public let from: String
        public let to: String
        public let kind: Kind
    }

    public struct Input: Sendable {
        public var focusId: String
        public var parentEdges: [(parent: String, child: String)]
        public var spouseEdges: [(a: String, b: String)]
        public var siblingEdges: [(a: String, b: String)]
        public var persons: [PersonAttributes]
        public var nodeWidth: Double
        public var nodeHeight: Double
        public var hGap: Double
        public var vGap: Double

        public init(
            focusId: String,
            parentEdges: [(parent: String, child: String)],
            spouseEdges: [(a: String, b: String)] = [],
            siblingEdges: [(a: String, b: String)] = [],
            persons: [PersonAttributes] = [],
            nodeWidth: Double = 132,
            nodeHeight: Double = 168,
            hGap: Double = 22,
            vGap: Double = 58
        ) {
            self.focusId = focusId
            self.parentEdges = parentEdges
            self.spouseEdges = spouseEdges
            self.siblingEdges = siblingEdges
            self.persons = persons
            self.nodeWidth = nodeWidth
            self.nodeHeight = nodeHeight
            self.hGap = hGap
            self.vGap = vGap
        }
    }

    public struct Result: Sendable {
        public let nodes: [Node]
        public let edges: [Edge]
        public let width: Double
        public let height: Double

        public func node(_ personId: String) -> Node? {
            nodes.first { $0.personId == personId }
        }
    }

    public static func compute(_ input: Input) -> Result {
        guard !input.focusId.isEmpty else {
            return Result(nodes: [], edges: [], width: 0, height: 0)
        }

        var sexOf: [String: String] = [:]
        var birthYearOf: [String: Int] = [:]
        for person in input.persons {
            sexOf[person.id] = person.sex ?? "unknown"
            if let year = person.birthYear { birthYearOf[person.id] = year }
        }

        var parentsOf: [String: [String]] = [:]
        var childrenOf: [String: [String]] = [:]
        for edge in input.parentEdges {
            childrenOf[edge.parent, default: []].append(edge.child)
            parentsOf[edge.child, default: []].append(edge.parent)
        }
        var spousesOf: [String: [String]] = [:]
        for edge in input.spouseEdges {
            spousesOf[edge.a, default: []].append(edge.b)
            spousesOf[edge.b, default: []].append(edge.a)
        }
        var siblingsOf: [String: [String]] = [:]
        for edge in input.siblingEdges {
            siblingsOf[edge.a, default: []].append(edge.b)
            siblingsOf[edge.b, default: []].append(edge.a)
        }

        // ── Generations: BFS up and down, nearest-to-focus wins, spouses share a row ──
        var generation: [String: Int] = [input.focusId: 0]
        var queue = [input.focusId]
        var head = 0
        while head < queue.count {
            let id = queue[head]; head += 1
            let g = generation[id]!
            for parent in parentsOf[id] ?? [] where generation[parent] == nil {
                generation[parent] = g - 1
                queue.append(parent)
            }
            for child in childrenOf[id] ?? [] where generation[child] == nil {
                generation[child] = g + 1
                queue.append(child)
            }
            for spouse in spousesOf[id] ?? [] where generation[spouse] == nil {
                generation[spouse] = g
                queue.append(spouse)
            }
            for sibling in siblingsOf[id] ?? [] where generation[sibling] == nil {
                generation[sibling] = g
                queue.append(sibling)
            }
        }
        let present = Set(generation.keys)

        // ── Branches: which side of the focus each ancestor comes from ───────────────
        let branchOf = assignBranches(
            focusId: input.focusId,
            parentsOf: parentsOf,
            spousesOf: spousesOf,
            sexOf: sexOf,
            present: present
        )

        // ── Couples: spouses, plus any two parents of the same child ────────────────
        var partners: [String: Set<String>] = [:]
        func link(_ a: String, _ b: String) {
            guard a != b, present.contains(a), present.contains(b),
                  generation[a] == generation[b] else { return }
            partners[a, default: []].insert(b)
            partners[b, default: []].insert(a)
        }
        for edge in input.spouseEdges { link(edge.a, edge.b) }
        for parents in parentsOf.values {
            for i in parents.indices {
                for j in parents.index(after: i)..<parents.endIndex { link(parents[i], parents[j]) }
            }
        }

        var byGeneration: [Int: [String]] = [:]
        for (id, g) in generation { byGeneration[g, default: []].append(id) }
        // Sorting the members keeps the layout deterministic; a dictionary's order is not.
        for key in byGeneration.keys { byGeneration[key]?.sort() }
        let generations = byGeneration.keys.sorted()

        func orderPair(_ x: String, _ y: String) -> [String] {
            let sx = sexOf[x], sy = sexOf[y]
            if sx == "male", sy == "female" { return [x, y] }
            if sx == "female", sy == "male" { return [y, x] }
            if let bx = birthYearOf[x], let by = birthYearOf[y], bx != by {
                return bx < by ? [x, y] : [y, x]
            }
            return x < y ? [x, y] : [y, x]
        }

        func orderComponent(_ members: [String]) -> [String] {
            if members.count <= 1 { return members }
            if members.count == 2 { return orderPair(members[0], members[1]) }
            // A chain or star: walk greedily from an endpoint.
            let set = Set(members)
            let start = members.first { (partners[$0] ?? []).filter(set.contains).count == 1 } ?? members[0]
            var sequence: [String] = []
            var seen: Set<String> = []
            var current: String? = start
            while let id = current {
                sequence.append(id)
                seen.insert(id)
                current = (partners[id] ?? []).sorted().first { set.contains($0) && !seen.contains($0) }
            }
            for member in members where !seen.contains(member) { sequence.append(member) }
            return sequence
        }

        var order: [String: Int] = [:]

        func branchRank(_ members: [String]) -> Int {
            let branches = Set(members.map { branchOf[$0] ?? .neutral })
            if branches == [.paternal] { return 0 }
            if branches == [.maternal] { return 2 }
            return 1
        }

        func orderGeneration(_ g: Int, neighbour: Int?) {
            guard let ids = byGeneration[g] else { return }

            // Union-find over the couple links inside this generation.
            var parent: [String: String] = [:]
            for id in ids { parent[id] = id }
            func find(_ x: String) -> String {
                var root = x
                while parent[root] != root { root = parent[root]! }
                return root
            }
            for id in ids {
                for other in partners[id] ?? [] where generation[other] == g {
                    parent[find(id)] = find(other)
                }
            }
            var components: [String: [String]] = [:]
            for id in ids { components[find(id), default: []].append(id) }

            func barycentre(_ id: String) -> Double {
                guard let neighbour else { return .greatestFiniteMagnitude }
                let related = g < 0 ? (childrenOf[id] ?? []) : (parentsOf[id] ?? [])
                let placed = related
                    .filter { generation[$0] == neighbour && order[$0] != nil }
                    .map { Double(order[$0]!) }
                guard !placed.isEmpty else { return .greatestFiniteMagnitude }
                return placed.reduce(0, +) / Double(placed.count)
            }

            let ordered = components.values
                .map { members -> (sequence: [String], bary: Double, branch: Int, key: String) in
                    let bary = members.map(barycentre).filter { $0 < .greatestFiniteMagnitude }
                    return (
                        orderComponent(members),
                        bary.isEmpty ? .greatestFiniteMagnitude : bary.reduce(0, +) / Double(bary.count),
                        branchRank(members),
                        members.min() ?? ""
                    )
                }
                // Paternall left, maternal right, everything else between — then by
                // barycentre so a couple sits under the kin it belongs to. The id is the
                // final tie-break, without which equal barycentres shuffle between runs.
                .sorted {
                    if $0.branch != $1.branch { return $0.branch < $1.branch }
                    if $0.bary != $1.bary { return $0.bary < $1.bary }
                    return $0.key < $1.key
                }

            var index = 0
            for component in ordered {
                for id in component.sequence {
                    order[id] = index
                    index += 1
                }
            }
        }

        orderGeneration(0, neighbour: nil)
        for g in generations.sorted(by: { abs($0) < abs($1) }) where g != 0 {
            orderGeneration(g, neighbour: g < 0 ? g + 1 : g - 1)
        }

        // ── Coordinates ─────────────────────────────────────────────────────────────
        let minGeneration = generations.first ?? 0
        let rowStep = input.nodeHeight + input.vGap
        let columnStep = input.nodeWidth + input.hGap
        let maxColumns = byGeneration.values.map(\.count).max() ?? 0

        var nodes: [Node] = []
        for g in generations {
            let ids = (byGeneration[g] ?? []).sorted { (order[$0] ?? 0) < (order[$1] ?? 0) }
            let rowOffset = Double(maxColumns - ids.count) * columnStep / 2
            for (column, id) in ids.enumerated() {
                nodes.append(Node(
                    personId: id,
                    generation: g,
                    x: rowOffset + Double(column) * columnStep,
                    y: Double(g - minGeneration) * rowStep,
                    coupleSide: .none,
                    branch: branchOf[id] ?? .neutral,
                    // Each generation away from the focus shifts the shade, so a chart with
                    // six generations reads as depth rather than as one flat blue.
                    tone: min(1, Double(abs(g)) * 0.18)
                ))
            }
        }

        // Which way a portrait should face: toward the nearest partner.
        let positions = Dictionary(uniqueKeysWithValues: nodes.map { ($0.personId, $0.x) })
        for index in nodes.indices {
            let candidates = (partners[nodes[index].personId] ?? []).compactMap { positions[$0] }
            guard let nearest = candidates.min(by: {
                abs($0 - nodes[index].x) < abs($1 - nodes[index].x)
            }) else { continue }
            nodes[index].coupleSide = nodes[index].x < nearest ? .left : .right
        }

        var edges: [Edge] = []
        for edge in input.parentEdges where present.contains(edge.parent) && present.contains(edge.child) {
            edges.append(Edge(from: edge.parent, to: edge.child, kind: .parent))
        }
        for edge in input.spouseEdges where present.contains(edge.a) && present.contains(edge.b) {
            edges.append(Edge(from: edge.a, to: edge.b, kind: .spouse))
        }
        for edge in input.siblingEdges
        where present.contains(edge.a) && present.contains(edge.b) && generation[edge.a] == generation[edge.b] {
            edges.append(Edge(from: edge.a, to: edge.b, kind: .sibling))
        }

        return Result(
            nodes: nodes,
            edges: edges,
            width: max(0, Double(maxColumns) * columnStep - input.hGap),
            height: max(0, Double(generations.count) * rowStep - input.vGap)
        )
    }

    /// Which side of the family each person sits on, seen from the focus.
    ///
    /// The father's ancestors are paternal and the mother's are maternal; a spouse takes the
    /// branch of the blood relative they married into. Everybody else — the focus, siblings,
    /// descendants — is neutral. This is what makes the chart symmetric: the two sides are a
    /// hard constraint on horizontal order, not a colour applied afterwards.
    static func assignBranches(
        focusId: String,
        parentsOf: [String: [String]],
        spousesOf: [String: [String]],
        sexOf: [String: String],
        present: Set<String>
    ) -> [String: Branch] {
        var branch: [String: Branch] = [focusId: .neutral]

        let parents = (parentsOf[focusId] ?? []).sorted()
        // Sex decides which parent starts which side. With no sex recorded the order is
        // stable rather than correct — better a consistent chart than one that flips.
        let father = parents.first { sexOf[$0] == "male" } ?? parents.first
        let mother = parents.first { sexOf[$0] == "female" && $0 != father } ?? parents.first { $0 != father }

        // Both seeds are claimed before either walk starts.
        //
        // The walks follow spouses — a person who married into the paternal line is paternal —
        // and the focus's own parents are spouses of each other, so whichever side went first
        // would otherwise swallow the other's founder and paint the whole chart one colour.
        if let father, present.contains(father) { branch[father] = .paternal }
        if let mother, present.contains(mother) { branch[mother] = .maternal }

        func walkUp(from start: String?, as side: Branch) {
            guard let start, present.contains(start) else { return }
            var queue = [start]
            var head = 0
            var expanded: Set<String> = []
            while head < queue.count {
                let id = queue[head]; head += 1
                // First assignment wins: a person reachable from both sides (pedigree
                // collapse — cousins who married) keeps the side that reached them first.
                // The seeds are already assigned, so they are expanded rather than skipped.
                if branch[id] == nil { branch[id] = side }
                else if branch[id] != side { continue }
                guard expanded.insert(id).inserted else { continue }
                for parent in (parentsOf[id] ?? []).sorted() where present.contains(parent) {
                    queue.append(parent)
                }
                for spouse in (spousesOf[id] ?? []).sorted() where present.contains(spouse) {
                    queue.append(spouse)
                }
            }
        }

        walkUp(from: father, as: .paternal)
        walkUp(from: mother, as: .maternal)
        return branch
    }
}
